src/ILLink.CodeFix/RequiresUnreferencedCodeCodeFixProvider.cs

```diff
@@ -5,120 +5,37 @@
 using System;
 using System.Collections.Immutable;
 using System.Composition;
-using System.Linq;
-using System.Threading;
 using System.Threading.Tasks;
 using ILLink.RoslynAnalyzer;
 using Microsoft.CodeAnalysis;
-using Microsoft.CodeAnalysis.CodeActions;
 using Microsoft.CodeAnalysis.CodeFixes;
 using Microsoft.CodeAnalysis.CSharp;
-using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Editing;
-using Microsoft.CodeAnalysis.Simplification;
 
 namespace ILLink.CodeFix
 {
 	[ExportCodeFixProvider (LanguageNames.CSharp, Name = nameof (RequiresUnreferencedCodeCodeFixProvider)), Shared]
-	public class RequiresUnreferencedCodeCodeFixProvider : CodeFixProvider
+	public class RequiresUnreferencedCodeCodeFixProvider : BaseAttributeCodeFixProvider
 	{
 		private const string s_title = "Add RequiresUnreferencedCode attribute to parent method";
 
 		public sealed override ImmutableArray<string> FixableDiagnosticIds
 			=> ImmutableArray.Create (RequiresUnreferencedCodeAnalyzer.DiagnosticId);
 
-		public sealed override FixAllProvider GetFixAllProvider ()
-		{
-			// See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/FixAllProvider.md for more information on Fix All Providers
-			return WellKnownFixAllProviders.BatchFixer;
-		}
-
 		public sealed override async Task RegisterCodeFixesAsync (CodeFixContext context)
 		{
-			var root = await context.Document.GetSyntaxRootAsync (context.CancellationToken).ConfigureAwait (false);
-
-			var diagnostic = context.Diagnostics.First ();
-			var diagnosticSpan = diagnostic.Location.SourceSpan;
-
-			// Find the containing method
-			SyntaxNode targetNode = root!.FindNode (diagnosticSpan);
-			CSharpSyntaxNode? declarationSyntax = null;
-			for (SyntaxNode? current = targetNode.Parent;
-				 current is not null;
-				 current = current.Parent) {
-				if (current is LambdaExpressionSyntax) {
-					return;
-				} else if (current.IsKind (SyntaxKind.LocalFunctionStatement)
-					  || current is BaseMethodDeclarationSyntax) {
-					declarationSyntax = (CSharpSyntaxNode) current;
-					break;
-				}
-			}
-
-			if (declarationSyntax is not null) {
-				var semanticModel = await context.Document
-					.GetSemanticModelAsync (context.CancellationToken).ConfigureAwait (false);
-				var symbol = semanticModel!.Compilation.GetTypeByMetadataName (
-					RequiresUnreferencedCodeAnalyzer.FullyQualifiedRequiresUnreferencedCodeAttribute);
-
-				// Register a code action that will invoke the fix.
-				context.RegisterCodeFix (
-					CodeAction.Create (
-						title: s_title,
-						createChangedDocument: c => AddRequiresUnreferencedCode (
-							context.Document, root, targetNode, declarationSyntax, symbol!, c),
-						equivalenceKey: s_title),
-					diagnostic);
-
-			}
+			await BaseRegisterCodeFixesAsync (context, AttributeableParentTargets.Method, RequiresUnreferencedCodeAnalyzer.FullyQualifiedRequiresUnreferencedCodeAttribute, s_title);
```

Comment:
Don't call `await` on Task-returning functions as the last statement, just return the result of the call and change the parent method to not be `async`.